pkg/kubelet/cm/dra/healthinfo.go

```diff
@@ -0,0 +1,195 @@
+	res := state.DeviceHealthStringUnknown
+	now := time.Now()
+
+	_ = cache.withRLock(func() error {
+		if driver, ok := (*cache.HealthInfo)[driverName]; ok {
+			for _, device := range driver.Devices {
+				if device.PoolName == poolName && device.DeviceName == deviceName {
+					if now.Sub(device.LastUpdated) > healthTimeout {
+						res = state.DeviceHealthStringUnknown
+					} else {
+						res = device.Health
+					}
+					return nil
+				}
+			}
+		}
+		return nil
+	})
+	return res
+}
+
+// updateHealthInfo updates the cache with a list of devices for driver, reconciling the full state.
+func (cache *healthInfoCache) updateHealthInfo(driverName string, devices []state.DeviceHealth) ([]state.DeviceHealth, bool, error) {
+	var changed bool
+	changedDevices := []state.DeviceHealth{}
+	err := cache.withLock(func() error {
+		now := time.Now()
+		currentDriver, exists := (*cache.HealthInfo)[driverName]
+		if !exists {
+			currentDriver = state.DriverHealthState{Devices: []state.DeviceHealth{}}
+		}
+
+		// Build map of reported devices
+		// Note: Within the driver's cache, devices can be keyed by '<pool name>/<device name>'
+		// as each update iteration is within one driver's scope.
+		reported := make(map[string]state.DeviceHealth)
+		for _, dev := range devices {
+			dev.LastUpdated = now
+			key := dev.PoolName + "/" + dev.DeviceName
+			reported[key] = dev
+		}
+
+		// Reconcile the cache: Update existing, add new, and mark unreported devices as "Unknown"
+		newDevices := make([]state.DeviceHealth, 0, len(reported))
+		for _, existing := range currentDriver.Devices {
+			key := existing.PoolName + "/" + existing.DeviceName
+			if updated, ok := reported[key]; ok {
+				if existing.Health != updated.Health {
+					changed = true
+					changedDevices = append(changedDevices, updated)
+				}
+				// Add  device to new device health status list
+				newDevices = append(newDevices, updated)
+				// Remove from reported device list
+				delete(reported, key)
+				// If device surpasses health timeout set to unknown
+			} else if existing.Health != state.DeviceHealthStringUnknown && now.Sub(existing.LastUpdated) > healthTimeout {
+				existing.Health = state.DeviceHealthStringUnknown
+				existing.LastUpdated = now
+				changed = true
+				changedDevices = append(changedDevices, existing)
+				newDevices = append(newDevices, existing)
+				// Within the health timeout consider healthy and add as new device
+			} else {
+				newDevices = append(newDevices, existing)
+			}
+		}
+
+		// Add remaining new devices
+		for _, dev := range reported {
+			dev.LastUpdated = now
+			newDevices = append(newDevices, dev)
+			changedDevices = append(changedDevices, dev)
+			changed = true
+		}
+
+		currentDriver.Devices = newDevices
+		(*cache.HealthInfo)[driverName] = currentDriver
+
+		if changed {
+			if err := cache.saveToCheckpointInternal(); err != nil {
+				klog.Background().Error(err, "Failed to save health checkpoint after update")
```

Comment:
Why is err not propagated for this case?

Reply:
nit: update log message to something that will explain what it means:

```suggestion
				klog.Background().Error(err, "Failed to save health checkpoint after update. Kubelet restart may lose the device health information.")
```